src/lib/components/Modal/Modal.jsx

```diff
@@ -78,54 +79,120 @@ export const Modal = ({
   primaryButtonRef,
   size,
   ...restProps
 }) => {
   const childrenWrapperRef = useRef();
 
-  const keyPressHandler = (e) => {
-    if (e.key === 'Escape' && closeButtonRef?.current != null) {
-      closeButtonRef.current.click();
+  useEffect(() => {
+    if (!autoFocus) {
+      return () => {};
     }
 
-    if (e.key === 'Enter' && e.target.nodeName !== 'BUTTON' && primaryButtonRef?.current != null) {
-      primaryButtonRef.current.click();
+    // If `autoFocus` is set to `true`, following code finds all focusable elements and among them
+    // first not disabled form field element (input, textarea or select) or primary button
+    // and auto focuses it. This is necessary to have focus on one of those elements to be able
+    // to submit the form by pressing Enter key. If there are neither, it tries to focus any other
+    // focusable elements. In case there are none, childrenWrapperElement (Modal itself) is focused.
+
+    if (childrenWrapperRef?.current == null) {
+      return () => {};
     }
-  };
 
-  useEffect(() => {
-    window.document.addEventListener('keydown', keyPressHandler, false);
-    const removeKeyPressHandler = () => {
-      window.document.removeEventListener('keydown', keyPressHandler, false);
-    };
+    const childrenWrapperElement = childrenWrapperRef.current;
+    const childrenFocusableElements = Array.from(
+      childrenWrapperElement.querySelectorAll('button, [href], input, select, textarea, [tabindex]:not([tabindex="-1"])'),
+    );
 
-    // If `autoFocus` is set to `true`, following code finds first form field element
-    // (input, textarea or select) or primary button and auto focuses it. This is necessary
-    // to have focus on one of those elements to be able to submit form by pressing Enter key.
-    if (autoFocus) {
-      if (childrenWrapperRef?.current != null) {
-        const childrenWrapperElement = childrenWrapperRef.current;
-        const childrenElements = childrenWrapperElement.querySelectorAll('*');
-        const formFieldEl = Array.from(childrenElements).find(
-          (element) => ['INPUT', 'TEXTAREA', 'SELECT'].includes(element.nodeName) && !element.disabled,
-        );
-
-        if (formFieldEl) {
-          formFieldEl.focus();
-          return removeKeyPressHandler;
-        }
+    const firstFormFieldEl = childrenFocusableElements.find(
+      (element) => ['INPUT', 'TEXTAREA', 'SELECT'].includes(element.nodeName) && !element.disabled,
+    );
+
+    if (childrenFocusableElements.length === 0) {
+      childrenWrapperElement.tabIndex = 1;
+      childrenFocusableElements.push(childrenWrapperElement);
+    }
+
+    const firstFocusableElement = childrenFocusableElements[0];
+    const lastFocusableElement = childrenFocusableElements[childrenFocusableElements.length - 1];
+
+    if (firstFormFieldEl) {
+      firstFormFieldEl.focus();
+    } else if (primaryButtonRef?.current != null) {
+      primaryButtonRef.current.focus({ focusVisible: true });
```

Comment:
The `focusVisible` option works only in FF. It would be nice to have similar behaviour (to indicate we moved focus in JS by showing outline - by default it does not show on buttons) in other browsers too. Any suggestions @adamkudrna ?

Reply:
> f you think the above use case is ok to leave up to the browser, I will remove the focus-visible option, because it would then work differently in chrome, edge and FF.

I agree.